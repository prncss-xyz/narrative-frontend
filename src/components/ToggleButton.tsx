import { Box } from "./basics";

/*
 * a simple toggling button
 * when setActive is undefined, the button is readonly
 */
export function ToggleButton({
  active,
  setActive,
  children,
  ...props
}: {
  active: boolean;
  setActive?: (b: boolean) => void;
  children: React.ReactNode;
  [prop: string]: unknown; // TODO: could be more restrictive
}) {
  return (
    <Box
      as="button"
      px={2}
      py={1}
      height={1}
      borderStyle="solid"
      borderColor={active ? "black" : "muted"}
      color={active ? "text" : "muted"}
      backgroundColor={active ? "accent" : "bgLight"}
      fontSize={0}
      borderRadius={3}
      onClick={setActive ? () => setActive(!active) : undefined}
      cursor={setActive ? undefined : "pointer"}
      {...props}
    >
      {children}
    </Box>
  );
}
