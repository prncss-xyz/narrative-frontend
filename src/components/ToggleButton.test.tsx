import { fireEvent, render, screen } from "@testing-library/react";
import { ToggleButton } from "./ToggleButton";

describe("ToggleButton", () => {
  describe("not disabled", () => {
    it("should trigger when clicked", () => {
      const spy = vi.fn();
      const { container } = render(
        <ToggleButton onClick={spy}>not disabled</ToggleButton>
      );
      fireEvent.click(screen.getByText("not disabled"));
      expect(spy).toHaveBeenCalledTimes(1);
      expect(container).toMatchSnapshot();
    });
  });
  describe("disabled", () => {
    it("should trigger when clicked", () => {
      const spy = vi.fn();
      const { container } = render(
        <ToggleButton disabled={true} onClick={spy}>
          disabled
        </ToggleButton>
      );
      fireEvent.click(screen.getByText("disabled"));
      expect(spy).toHaveBeenCalledTimes(0);
      expect(container).toMatchSnapshot();
    });
  });
});
