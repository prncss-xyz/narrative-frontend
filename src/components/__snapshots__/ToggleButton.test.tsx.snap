// Vitest Snapshot v1

exports[`ToggleButton > disabled > should trigger when clicked 1`] = `
<div>
  <button
    class="css-1717d9d-Box e2i8k6n3"
    color="text"
    cursor="pointer"
    disabled=""
    font-size="0"
    height="1"
  >
    disabled
  </button>
</div>
`;

exports[`ToggleButton > not disabled > should trigger when clicked 1`] = `
<div>
  <button
    class="css-1gxxcqd-Box e2i8k6n3"
    color="muted"
    font-size="0"
    height="1"
  >
    not disabled
  </button>
</div>
`;
